api/dataset_api.go

```diff
@@ -53,94 +54,108 @@ type JobLinks struct {
-	jsonBody, httpCode, err := api.get(path, nil)
+	jsonBody, httpCode, err := api.get(ctx, path, nil)
 	logData["httpCode"] = httpCode
-	if httpCode == http.StatusNotFound {
-		return Instance{}, nil
-	}
 	if err == nil && httpCode != http.StatusOK {
 		err = errors.New("Bad response while getting dataset instance")
+		if httpCode < http.StatusInternalServerError {
+			isFatal = true
+		}
+	} else if err != nil {
+		isFatal = true
 	}
 	if err != nil {
 		log.ErrorC("GetInstance get", err, logData)
-		return Instance{}, err
+		return
 	}
 	logData["jsonBody"] = jsonBody
 
-	var instance Instance
-	if err := json.Unmarshal(jsonBody, &instance); err != nil {
+	if err = json.Unmarshal(jsonBody, &instance); err != nil {
 		log.ErrorC("GetInstance unmarshall", err, logData)
-		return Instance{}, err
+		isFatal = true
 	}
-	return instance, nil
+	return
 }
 
 // GetInstances asks the Dataset API for all instances filtered by vars
-func (api *DatasetAPI) GetInstances(vars url.Values) ([]Instance, error) {
+func (api *DatasetAPI) GetInstances(ctx context.Context, vars url.Values) (instances []Instance, err error, isFatal bool) {
 	path := api.url + "/instances"
 	logData := log.Data{"path": path}
-	jsonBody, httpCode, err := api.get(path, vars)
+	jsonBody, httpCode, err := api.get(ctx, path, vars)
 	logData["httpCode"] = httpCode
 	if err == nil && httpCode != http.StatusOK {
 		err = errors.New("Bad response while getting dataset instances")
+		if httpCode < http.StatusInternalServerError {
+			isFatal = true
+		}
+	} else if err != nil {
+		isFatal = true
 	}
 	if err != nil {
 		log.ErrorC("GetInstances get", err, logData)
-		return nil, err
+		return
 	}
 	logData["jsonBody"] = jsonBody
 
 	var instanceResults InstanceResults
-	if err := json.Unmarshal(jsonBody, &instanceResults); err != nil {
+	if err = json.Unmarshal(jsonBody, &instanceResults); err != nil {
 		log.ErrorC("GetInstances Unmarshal", err, logData)
-		return nil, err
+		return instances, err, true
 	}
-	return instanceResults.Items, nil
+	return instanceResults.Items, nil, isFatal
 }
 
 // UpdateInstanceWithNewInserts tells the Dataset API of a number of observationsInserted for instanceID
-func (api *DatasetAPI) UpdateInstanceWithNewInserts(instanceID string, observationsInserted int32) error {
+func (api *DatasetAPI) UpdateInstanceWithNewInserts(ctx context.Context, instanceID string, observationsInserted int32) (err error, isFatal bool) {
 	path := api.url + "/instances/" + instanceID + "/inserted_observations/" + strconv.FormatInt(int64(observationsInserted), 10)
 	logData := log.Data{"url": path}
-	jsonBody, httpCode, err := api.put(path, nil)
+	jsonBody, httpCode, err := api.put(ctx, path, nil)
 	logData["httpCode"] = httpCode
 	logData["jsonBytes"] = jsonBody
 	if err == nil && httpCode != http.StatusOK {
 		err = errors.New("Bad response while updating inserts for job")
+		if httpCode < http.StatusInternalServerError {
```

Comment:
ditto

Reply:
I've moved all this error checking to its own function - too much repetition